File: tree/BinaryTreeTraversal_BFS_Use_Stack.py
'''
Created on 2018年8月4日
https://www.geeksforgeeks.org/inorder-tree-traversal-without-recursion/
@author: Viking
'''

# python3 program to do inorder traversal without recursion
# by use Stack


class Node:
    def __init__(self, data):
        self.data = data
        self.left = self.right = None


def Inorder(root):
    stack = []
    current = root
    while ((current is not None) or (len(stack) > 0)):
        while current is not None:
            stack.append(current)
            current = current.left

        current = stack.pop()
        print(current.data, end=" ")
        current = current.right

# Driver program to test above function


""" Constructed binary tree is
            1
          /   \
         2     3
       /  \
      4    5   """

root = Node(1)
root.left = Node(2)
root.right = Node(3)
root.left.left = Node(4)
root.left.right = Node(5)

Inorder(root)


'''
4 2 5 1 3 
'''

File: tree/BinaryTreeTraversal_BFS_Use_Stack.java
package org.q10viking.tree;

import java.util.Stack;
class SNode{
	int data;
	SNode left,right;
	public SNode(int data) {
		this.data = data;
		this.left = this.right = null;
	}
}
public class BinaryTreeTraversal_BFS_Use_Stack {
	SNode root;
	BinaryTreeTraversal_BFS_Use_Stack(SNode root){
		this.root = root;
	}
	
	void Inorder() {
		Stack<SNode> stack = new Stack<SNode>();
		SNode current = this.root;
		while(current != null || stack.size()>0) {
			
			while(current != null) {
				/*place pointer to a tree SNode on the 
				 * stack before traversing the SNode's left 
				 */
				stack.push(current);
				current = current.left;
			}
			
			//current must be null at this point
			current = stack.pop();
			System.out.print(current.data+" ");
			current = current.right;
			
		
		}
		
	}
	
	public static void main(String[] args) {
		SNode root;
		root = new SNode(1);
        root.left = new SNode(2);
        root.right = new SNode(3);
        root.left.left = new SNode(4);
        root.left.right = new SNode(5);
        
        BinaryTreeTraversal_BFS_Use_Stack tree = new BinaryTreeTraversal_BFS_Use_Stack(root);
        tree.Inorder();
        
	}

}
/*
4 2 5 1 3 

*/
